### PR TITLE
[FC-0056][Plugin] Course outline sidebar

### PR DESCRIPTION
**Pull request by ihor-romaniuk**
**Settings**

```yaml
EDX_PLATFORM_REPOSITORY: https://github.com/raccoongang/edx-platform.git
EDX_PLATFORM_VERSION: ts-develop
```

### Description
This pull request adds an important feature to our platform: displaying a navigation sidebar within a given course.

**Note:** Initial solution without using `Frontend Plugin Framework` https://github.com/openedx/frontend-app-learning/pull/1342 

### Design
https://www.figma.com/file/gew5tORDX4Q7wxOS8vjqZu/side-nav-OEX?type=design&node-id=318-3234&mode=design&t=rBe1ToNYP8RY6QOp-0

image

### Testing instructions
1. Run master devstack.
2. Start platform make dev.up.lms+cms+frontend-app-course-authoring and make checkout on this branch.
3. Go to Course Unit page from the Course Outline page.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @ihor-romaniuk! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

File: src/courseware/course/Course.jsx [hunk 1]
```diff
@@ -37,7 +36,8 @@ const Course = ({
   const sequence = useModel('sequences', sequenceId);
   const section = useModel('sections', sequence ? sequence.sectionId : null);
   const enableNewSidebar = getConfig().ENABLE_NEW_SIDEBAR;
-  const navigationDisabled = sequence?.navigationDisabled ?? false;
+  const { enabled: isEnabledOutlineSidebar = false } = useSelector(getCoursewareOutlineSidebarSettings);
+  const navigationDisabled = isEnabledOutlineSidebar || (sequence?.navigationDisabled ?? false);
```
Comment (ihor-romaniuk) [inline src/courseware/course/Course.jsx:40]: The `useSelector(getCoursewareOutlineSidebarSettings)` should be replaced with a method that check if the Sidebar plugin is available and displayed.

File: src/courseware/course/sidebar/sidebars/index.js [hunk 2]
```diff
@@ -6,15 +6,25 @@ export const SIDEBARS = {
...
   notifications.ID,
 ];
+
+export const COURSE_OUTLINE_SIDEBAR_ID = 'COURSE_OUTLINE_SIDEBAR';
+
+export const extendSidebars = (key, value) => {
+  SIDEBARS[key] = value;
+};
+
+export const checkIsSidebarAvailable = (id) => Object.keys(SIDEBARS).includes(id);
```
Comment (ihor-romaniuk) [inline src/courseware/course/sidebar/sidebars/index.js:30]: Through the close interaction of the `Outline Sidebar` plugin's display with the application's functionality display (sidebar display order, breadcrumb display, etc.), we should have a way to check for an active `Outline Sidebar` plugin.

Perhaps it should be reworked and placed in a global state and have a list of connected plugins.

File: src/index.scss [hunk 3]
```diff
@@ -400,10 +404,14 @@
...
 @import "course-home/progress-tab/course-completion/CompletionDonutChart.scss";
 @import "course-home/progress-tab/grades/course-grade/GradeBar.scss";
 @import "courseware/course/course-exit/CourseRecommendations";
 @import "product-tours/newUserCourseHomeTour/NewUserCourseHomeTourModal.scss";
 @import "course-home/courseware-search/courseware-search.scss";
 @import "course-tabs/course-tabs-navigation.scss";
+@import "courseware/course/sidebar/common/SidebarBase.scss";
+
+// Import plugin-specific sass files
+@import "../plugins/index.scss";
```
Comment (ihor-romaniuk) [inline src/index.scss:417]: We needed to be able to include style files from plugins with the ability to use the Paragon system, so we decided to add one entry point for all plugin styles.

**Review by arbrandes — CHANGES_REQUESTED** [review]
Thank you for taking on this work!  My first big request for a change, however, is the following:

Any feature that should be on by default - which is the case with this sidebar - should **not** be a plugin.  It can be implemented as default content inside a plugin slot, but users should not have to change their configuration to get the corresponding functionality.

In other words, I would like to see all components and corresponding data functions moved out of plugins, but with one caveat: we should create an example plugin that disables the functionality being introduced, and document how to use it.  This means we'll also have to test that this "disable plugin" actually works.

**Comment by ihor-romaniuk** [comment]
@arbrandes I created a new PR https://github.com/openedx/frontend-app-learning/pull/1349, which will wrap insertion points of triggers for calling the navigation sidebar, added sidebar accessibility checks to hide breadcrumbs when a sidebar is present, and added an example of a standard configuration in `env.config.jsx`.

**Comment by itsjeyd** [comment]
@ihor-romaniuk 

> I created a new PR https://github.com/openedx/frontend-app-learning/pull/1349, ...

Does that PR replace this one? Asking because I'd like to understand if we need to keep this PR open.

**Comment by itsjeyd** [comment]
OSPR management note: This is part of a set of PRs for implementing Phase 1 changes for the FC-0056 project. We can [consider it approved](https://github.com/openedx/frontend-app-learning/pull/1342#issuecomment-2074998975) from the product perspective.

CC @GlugovGrGlib

**Comment by openedx-webhooks** [comment]
@ihor-romaniuk Even though your pull request wasn’t merged, please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/frontend-app-learning&entry.867055334=https://github.com/openedx/frontend-app-learning/pull/1344&entry.1484655318=https://github.com/ihor-romaniuk&entry.752974735=2024-04-04+11:53&entry.1917517419=2024-04-26+09:19&entry.2133058324=No) so we can improve your experience in the future.

**Comment by ihor-romaniuk** [comment]
The main PR about the course outline sidebar functionality https://github.com/openedx/frontend-app-learning/pull/1349

**Pull request closed without merging**